// @keus-automation/kiotp-realtime-js/dist/src/nats-export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONCodec = exports.nkeyAuthenticator = exports.Empty = exports.jwtAuthenticator = exports.connect = exports.StringCodec = exports.Events = void 0;
const nats_1 = require("nats");
Object.defineProperty(exports, "StringCodec", { enumerable: true, get: function () { return nats_1.StringCodec; } });
Object.defineProperty(exports, "jwtAuthenticator", { enumerable: true, get: function () { return nats_1.jwtAuthenticator; } });
Object.defineProperty(exports, "connect", { enumerable: true, get: function () { return nats_1.connect; } });
Object.defineProperty(exports, "Empty", { enumerable: true, get: function () { return nats_1.Empty; } });
Object.defineProperty(exports, "nkeyAuthenticator", { enumerable: true, get: function () { return nats_1.nkeyAuthenticator; } });
Object.defineProperty(exports, "JSONCodec", { enumerable: true, get: function () { return nats_1.JSONCodec; } });
Object.defineProperty(exports, "Events", { enumerable: true, get: function () { return nats_1.Events; } });
