// @keus-automation/kiotp-realtime-js/dist/src/realtime-interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IKIOTPRealtimeCommunication = exports.Errors = exports.ConnectionEvents = void 0;
const events_1 = require("events");
exports.ConnectionEvents = {
    CONNECTING: 'Connecting',
    RECONNECTING: 'Reconnecting',
    CONNECTED: 'Connected',
    ERROR: 'Error'
};
var Errors;
(function (Errors) {
    class InvalidConnection extends Error {
        constructor() {
            super("Invalid Connection");
        }
        static getErrorCode() {
            return 301;
        }
    }
    Errors.InvalidConnection = InvalidConnection;
    class RPCNotAvailable extends Error {
        constructor() {
            super("RPC Not Available");
        }
        static getErrorCode() {
            return 302;
        }
    }
    Errors.RPCNotAvailable = RPCNotAvailable;
    class InvalidDataFormat extends Error {
        constructor() {
            super("Invalid Request or Response Data");
        }
        static getErrorCode() {
            return 303;
        }
    }
    Errors.InvalidDataFormat = InvalidDataFormat;
})(Errors = exports.Errors || (exports.Errors = {}));
class IKIOTPRealtimeCommunication extends events_1.EventEmitter {
    constructor(config) {
        super();
    }
    ;
}
exports.IKIOTPRealtimeCommunication = IKIOTPRealtimeCommunication;
